src/Storages/StorageS3.cpp

```diff
@@ -322,6 +339,161 @@ class StorageS3BlockOutputStream : public IBlockOutputStream
+            validateParitionKey(value);
+            auto [it, inserted] = sub_blocks_indices.emplace(value, sub_blocks_indices.size());
+            selector.push_back(it->second);
+        }
+
+        Blocks sub_blocks(sub_blocks_indices.size(), block.cloneEmpty());
+        for (size_t column_index = 0; column_index < block.columns(); ++column_index)
+        {
+            const auto & column = block.getByPosition(column_index);
+            MutableColumns sub_columns = column.column->scatter(sub_blocks.size(), selector);
+            for (size_t sub_column = 0; sub_column < sub_columns.size(); ++sub_column)
+            {
+                sub_blocks[sub_column].getByPosition(column_index).column = std::move(sub_columns[sub_column]);
+            }
+        }
+
+        for (const auto & [partition_id, index] : sub_blocks_indices)
+        {
+            getWriterForPartition(partition_id)->write(sub_blocks[index]);
+        }
+    }
+
+    void writePrefix() override
+    {
+        /// `writePrefix()` for each child writer is called in `getWriterForPartition()`
+        /// on writer creation.
+    }
+
+    void flush() override
+    {
+        for (auto & [partition_id, writer] : writers)
+        {
+            writer->flush();
+        }
+    }
+
+    void writeSuffix() override
+    {
+        for (auto & [partition_id, writer] : writers)
+        {
+            writer->writeSuffix();
+        }
+    }
+
+private:
+    const String format;
+    const Block sample_block;
+    ContextPtr context;
+    const CompressionMethod compression_method;
+    std::shared_ptr<Aws::S3::S3Client> client;
+    const String bucket;
+    const String key;
+    size_t min_upload_part_size;
+    size_t max_single_part_upload_size;
+
+    ExpressionActionsPtr partition_by_expr;
+    String partition_by_column_name;
+
+    std::unordered_map<String, BlockOutputStreamPtr> writers;
+
+    static String replaceWildcards(const String & input, const String & partition_id)
+    {
+        return boost::replace_all_copy(input, PARTITION_ID_WILDCARD, partition_id);
+    }
+
+    BlockOutputStreamPtr getWriterForPartition(const String & partition_id)
+    {
+        if (writers.count(partition_id) == 0)
+        {
+            writers.emplace(partition_id, std::make_shared<StorageS3BlockOutputStream>(
+                format,
+                sample_block,
+                context,
+                compression_method,
+                client,
+                replaceWildcards(bucket, partition_id),
+                replaceWildcards(key, partition_id),
+                min_upload_part_size,
+                max_single_part_upload_size
+            )).first->second->writePrefix();
+        }
+
+        return writers[partition_id];
+    }
+
+    static void validateParitionKey(const StringRef & str)
+    {
+        if (str.size == 0)
+            throw DB::Exception(ErrorCodes::CANNOT_PARSE_TEXT, "Illegal empty partition key");
```

Comment:
Why partition key can not be empty? Hypothetically, it can be a "sorted comma-separated list of binary features".

Reply:
If we allow empty keys such case produce unclear error:


`insert into table function s3('http://localhost:9001/foo/{_partition_id}', 'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') PARTITION BY val values (1, '');`

```
Received exception from server (version 21.9.1):
Code: 499. DB::Exception: Received from localhost:9000. DB::Exception: Unable to parse ExceptionName: MalformedXML Message: The XML you provided was not well-formed or did not validate against our published schema.. 
```

Reply:
Same in this case:
```
cristal :) insert into table function s3('http://localhost:9001/foo/', 'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') values (1, '');

INSERT INTO FUNCTION s3('http://localhost:9001/foo/', 'minioadmin', 'minioadmin', 'CSV', 'id Int32, val String') VALUES

Query id: f4c92770-b6fc-45f2-a6a6-5e8af62126b4


1 rows in set. Elapsed: 0.037 sec. 

Received exception from server (version 21.9.1):
Code: 499. DB::Exception: Received from localhost:9000. DB::Exception: Unable to parse ExceptionName: MalformedXML Message: The XML you provided was not well-formed or did not validate against our published schema.. 
```

I think we shall remove any validation in this pull request, BUT implement it in another, in `WriteBufferFromS3`/`ReadBufferFromS3`.